File: javascript/sqli_vulnerable.js
/**
 * Vulnerable JavaScript code with SQL injection flaws.
 * For security scanner testing only.
 */

const mysql = require('mysql');
const sqlite3 = require('sqlite3');

function getUserByIdVulnerable(userId) {
    // SQL injection via template literal
    const db = new sqlite3.Database('database.db');
    db.get(`SELECT * FROM users WHERE id = ${userId}`, (err, row) => {
        console.log(row);
    });
}

function searchUsersVulnerable(name) {
    // SQL injection via string concatenation
    const connection = mysql.createConnection({
        host: 'localhost',
        user: 'root',
        password: 'password',
        database: 'mydb'
    });

    const query = "SELECT * FROM users WHERE name = '" + name + "'";
    connection.query(query, (error, results) => {
        console.log(results);
    });
}

function getProductsVulnerable(category) {
    // SQL injection via template literal
    const connection = mysql.createConnection({});
    connection.query(`SELECT * FROM products WHERE category = '${category}'`, (err, results) => {
        console.log(results);
    });
}

function findByEmailVulnerable(email) {
    // SQL injection via string concatenation
    const db = new sqlite3.Database('database.db');
    const query = "SELECT * FROM users WHERE email = '" + email + "'";
    db.all(query, (err, rows) => {
        console.log(rows);
    });
}

function getUserSafe(userId) {
    // Safe parameterized query
    const connection = mysql.createConnection({});
    connection.query('SELECT * FROM users WHERE id = ?', [userId], (err, results) => {
        console.log(results);
    });
}

File: python/sqli_vulnerable.py
"""
Vulnerable Python code with SQL injection flaws.
For security scanner testing only.
"""
import sqlite3
import mysql.connector


def get_user_by_id_vulnerable(user_id):
    """SQL injection via f-string."""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM users WHERE id = {user_id}")
    return cursor.fetchone()


def search_users_vulnerable(name):
    """SQL injection via string concatenation."""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE name = '" + name + "'"
    cursor.execute(query)
    return cursor.fetchall()


def get_products_vulnerable(category):
    """SQL injection via string formatting."""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM products WHERE category = '%s'" % category)
    return cursor.fetchall()


def find_by_email_vulnerable(email):
    """SQL injection via .format()."""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE email = '{}'".format(email)
    cursor.execute(query)
    return cursor.fetchone()


def mysql_vulnerable(username, password):
    """MySQL SQL injection."""
    conn = mysql.connector.connect(host='localhost', user='root', password='pass')
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM users WHERE username='{username}' AND password='{password}'")
    return cursor.fetchone()


def get_user_safe(user_id):
    """Safe parameterized query."""
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))
    return cursor.fetchone()
